src/compiler/irgen.jl

```diff
@@ -221,6 +190,75 @@ function irgen(ctx::CompilerContext)
     return mod, entry
 end
 
+# this pass lowers `jl_throw` and friends to GPU-compatible exceptions.
+# this isn't strictly necessary, but has a couple of advantages:
+# - we can kill off unused exception arguments that otherwise would allocate or invoke
+# - we can fake debug information (lacking a stack unwinder)
+#
+# once we have thorough inference (ie. discarding `@nospecialize` and thus supporting
+# exception arguments) and proper debug info to unwind the stack, this pass can go.
+function lower_throw!(mod::LLVM.Module)
+    ctx = global_ctx::CompilerContext
+    changed = false
+
+    throw_functions = Dict{String,String}(
+        "jl_throw"                      => "exception",
+        "jl_error"                      => "error",
+        "jl_too_few_args"               => "too few arguments exception",
+        "jl_too_many_args"              => "too many arguments exception",
+        "jl_type_error_rt"              => "type error",
+        "jl_undefined_var_error"        => "undefined variable error",
+        "jl_bounds_error"               => "bounds error",
+        "jl_bounds_error_v"             => "bounds error",
+        "jl_bounds_error_int"           => "bounds error",
+        "jl_bounds_error_tuple_int"     => "bounds error",
+        "jl_bounds_error_unboxed_int"   => "bounds error",
+        "jl_bounds_error_ints"          => "bounds error",
+        "jl_eof_error"                  => "EOF error"
+    )
+
+    for (fn, name) in throw_functions
+        if haskey(functions(mod), fn)
+            f = functions(mod)[fn]
+
+            for use in uses(f)
+                call = user(use)::LLVM.CallInst
+
+                # replace the throw with a PTX-compatible exception
+                let builder = Builder(JuliaContext())
+                    position!(builder, call)
+                    emit_exception!(builder, name, call)
+                    dispose(builder)
+                end
+
+                # remove the call
+                call_args = collect(operands(call))[1:end-1] # last arg is function itself
+                unsafe_delete!(LLVM.parent(call), call)
+
+                # HACK: kill the exceptions' unused arguments
```

Comment:
DCE or ADCE is not enough?

Reply:
Sadly not, these are calls to `@nospecialize` constructors that contain allocations, or even calls to `jl_generic_apply`. LLVM doesn't touch those.

Reply:
Note that this hack used to be there in the past as well, it is now just better confined to a single pass.